### DISCUSSION TITLE
Mobile Push methods other than DUO?

### DISCUSSION THREAD

**Discussion by crogialli** [authelia/authelia#5337] [Ideas]
First of all, my best compliments for your work, it's **GREAT**.
I was wondering if the compulsory association of Mobile Push methodology with DUO is due to an **UNDERSTANDABLE** policy reason, or if we may hope to deliver Mobile Push notifications in different manners in the future.

As an example, in my environment I already implement an inexpensive GSM/3G/4G modem system capable of sending massive amounts of SMSs for free, if sent within my home country numbers (and that's the case in my application).
So, if a simple API capable to trigger a script once the the push notification is generated (reporting both the push notification code and any mean to recognize the user, such as the username or the e-mail address), would allow the notification delivery through the existing infrastructure.

I understand that the user's telephone number is something that's not maintained within the current authelia database, but it wouldn't be a great deal to retrieve it from LDAP or from am external db.

I've searched for similar discussions, but I wasn't able to find anyone, sorry if my question is a duplicate.

**Comment by james-d-elliott** [comment]
Push notifications from other means are definitely possible in the future, it's likely that we'll develop something ourselves at some stage.

I'm not sure we want to open the can of worms of SMS specifically however due to the massive security issues with that technology for 2FA, but maybe if the admin opts in with informed consent. If we did it we'd probably do it via sending an email to a recipient with a format  that includes the mobile number (i.e. number 1234560987 and format  `{{ .Number }}.sms@example.org` would trigger a send of the plaintext SMS content to `1234560987.sms@example.org` which then would be sent by the third party on to the SMS network). This methodology is incredibly common.

Critical Good Security Practice Implementation Elements:

- Make the availability opt-in by the administrator:
    - Ensure this cannot accidentally enabled by only documenting this feature in the official configuration documentation, and potentially specifically license it to require third parties link to the documentation and include all of the *up-to-date* preamble about these security implications.
    - Ensure there is very detailed information about this feature being discouraged due to security attack vectors beyond the control of Authelia, the administrator, or the end-user.
    - Ensure there is clear information about the more security friendly options available.
- Make the administrator provide the mobile number as part of the authentication backend (LDAP or file).
- Make the end-user opt-in to the feature by confirming their recorded number with:
    - Them entering the on file number (which must match).
    - Them receiving an initial SMS to confirm their possession of the device.
- Make sure the end-user is notified of:
    - Registration of this feature.
    - Usage of this feature (opt-out available if enabled by the administrator).

**Reply by crogialli** [reply]
Thank you for the answer, I'm looking forward to see those new features, hope you will be able to fit them in your development plan. I believe the "informed consent" is the right way ahead, since they are a "can of worm" as you say for general, undiscriminated use, but they are invaluable in some environments (zero need to setup anything at the client's side, zero need to have users registering in a service, very often they are also zero cost).

Thanks again, you're doing great.

**Reply by james-d-elliott** [reply]
No worries at all. I also had the idea that the end-user should be required to opt-in as well via the UI by confirming their mobile number (it will have to match what the administrator has). This confirmation will have to be recorded in the database and if the number recorded (in LDAP or the database file) by the administrator changes then this confirmation will have to occur again. This alleviates part of the numerous attack vectors that exist when using this method. 

I am still apprehensive about it being a good idea. But I will keep it in mind and we can see from the responses and upvotes here as to how much interest there is in such a feature, as well as consider it carefully and how to properly prevent potential issues (for example emailing a user at the same time to confirm this event occurred with an opt-out for this notification available).

**Reply by james-d-elliott** [reply]
I also updated the original post, the bottom list was added. See what you think.

**Comment by crogialli** [comment]
Thank you for your courtesy and attention; your to-do list is perfect.

Just about the opportunity of going this way or not, I understand that you are mainly concerned about the potential insecurity of some cloud-based (or simply internet-based) SMS providers.
But let me make you consider another scenario I tried to depict in the attached image, where **BOTH** authelia and SMS sending capabilities are **SELF-HOSTED**.

In that case, the potentially insecure link (the one in red) lies within the company network, under the total responsibility of the end user, and as safe as the company network in itself.

Why do I think authelia authors should care? Because one of the goals reached by authelia so far is being self-hosted. So, unless authelia authors decide to federate with one or more SMS providers for righteous business considerations, I believe the door towards a self-hosted solution to send OTPs shall also be left open.
Provided that, as you correctly observed, the end-user takes full responsibility for the security of the SMS transport.

I let the discussion to the team; your work is excellent regardless you decide to implement this or not.

<img width="1521" alt="Screenshot 2023-05-01 alle 19 39 42" src="https://user-images.githubusercontent.com/4084123/235501613-06841502-ac5d-45bd-a8ce-05e726f6273d.png">

**Reply by james-d-elliott** [reply]
The insecure portion is the cellular network. To those in the the security profession it's not considered a viable 2FA anymore, and probably never will be again by anyone serious about security, there are and have been too many methods to intercept the SMS itself before it reaches the users phone. Prime example was the Reddit breach where they gained control of Reddit Admin accounts (not the subredit admins, the actual admins). They were using SMS 2FA. The attackers received the SMS's on their cloned phones. This is getting harder to do. But not in all countries, and it's definitely still a very viable attack vector in most countries (it's gone from a very simple process to a slightly more complex one).

Given that an attacker knows the number of the user (which is a significantly easier to obtain via phishing than a password) for security purposes it should not be even considered a barrier.

We care because we (Authelia), the user, and the admin have no way to be in control of the security of these networks and they are far too often easily compromised. While the same security flaws can exist in other networks the cellular networks themselves are never held accountable for these issues, it's always the app which used them. In addition the key difference between them and someone like Duo, is that Duo is designed for this purpose, whereas solutions like SMS and Email are primarily communication systems. The issue with SMS when compared to email, is that there is no way for any end user or small company to secure SMS, whereas several measures can be implemented with email to secure it.

We could just not care and let users fall into this issue and throw our arms up in the air like everyone else and say "well not our problem, users should just know that this can happen". But this attitude is arguably the cause for all of these issues in the first place. Everyone is quick to avoid taking responsibility and  prioritizes features or "convenience" over properly implementing practical and secure solutions.

I'm not concerned about cloud SMS gateways anymore than self-hosted. In fact I'd consider the self-hosted or free cloud solutions less secure by and large than the premium solutions in this area in most instances. Mainly because the premium solutions have a chance to be audited. This however doesn't have any impact on the security of the cellular network itself.

**Reply by james-d-elliott** [reply]
For reference, I'm open to it, just feel it's warranted that we both gauge user interest ensuring it's wanted by more than just a handful of users and also take our time to consider the most appropriate way to secure something like this.

**Reply by crogialli** [reply]
Thank you for the explanation, I better understand your point of view.

**Comment by Smiley-k** [comment]
damn, I'm blind forgive me ..... I just created a similar request , but with a request for other code delivery tools that could have a demand from a large audience

https://github.com/authelia/authelia/discussions/6105

**Comment by nicerloop** [comment]
How about using WebPush notifications? These solely rely on the browser infrastructure to convey notifications from server to browser, and can be used to provide an URL to the Authelia portal to fill the second factor authentication (TOTP or WebAuthn).

Just created another tread about WebPush, as this thread seems to be centered on SMS notifications.

**Comment by james-d-elliott** [comment]
Technically the SMS element can already be satisfied today using email notifications and modifications to the text templates. You just have to have an intermediary service that accepts emails and implements the protocol related to your SMS gateways.

Since the original message was mostly surrounding this I'm closing it off.
